### ISSUE TITLE
Recreating k-means clustering order

### ISSUE THREAD

**Issue by AlexZambon** [jokergoo/ComplexHeatmap#737]
Hello, 
I'm trying to extract kmeans cluster transcripts. I used the following commands

Heatmap(as.matrix(Ht2236[,1:12]), 
        show_row_names = F, 
        km = 8) -> Hm2236
draw(Hm2236)

Why is it that every time I run only the line of code with the draw function i.e. draw(Hm2236) the K means cluster in the plot has different labels? Should the clusters be statically stored in the Hm2236 object? Otherwise if I extract the cluster gene groups and run analysis on them I can't be sure of what each cluster pattern is associated with which cluster group... I can understand why clusters would be different everytime the Heatmap argument is run due to organization. Alternatively can I set some type of sed function so that I get consistent results in terms of cluster outputs every time? Thank you!!!!

**Comment by jokergoo** [comment]
Because k-means has random start points. You can just set a fixed random seed:

```r
set.seed(123)
Your code
```

**Comment by AlexZambon** [comment]
Thank you for the quick response jokergoo! Just for my understanding and to be absolutely clear when you run the following line of code:

draw(Hm2236) 

This actually reclusters the data? I naively assumed that if I store the output of Heatmap() into an object (eg. Hm2236) it would store the k-means cluster information in that object (i.e. Hm2236). This does not appear to be correct, after testing the set.seed(123, as you suggested, I realized that I need to run the set.seed(123) line just before the draw() function otherwise I would get the different clusters each time. So to be consistent with pulling gene information out of the Heatmap() output and then matching those genes with the image output from draw() (for a publication for example I want to show the cluster and then gene ontology enrichment on clusters and then link those terms to specific clusters) I need to set.seed(123) twice and the code will need to look like the following:

set.seed(123)
Heatmap(as.matrix(Ht2236[,1:12]), show_row_names = F, km = 8) -> Hm2236
#then use a function to extract genes from each cluster here
set.seed(123)
draw(Hm2236) #this image will then match the genes from the clusters stored in Hm2236

Can you please confirm that this is correct? Thanks for such a great tool!

**Comment by jokergoo** [comment]
When you do this way:

```r
Heatmap(as.matrix(Ht2236[,1:12]),
    show_row_names = F,
    km = 8) -> Hm2236
draw(Hm2236)
```

Only heatmap-related data is stored in `Hm2236` while clustering is not applied yet. That is way every time when you run` draw()`, you get different k-means results.

You can do like follows:

```r
Heatmap(as.matrix(Ht2236[,1:12]),
    show_row_names = F,
    km = 8) -> Hm2236
Hm2236 = draw(Hm2236) # or Hm2236 = make_layout(Hm2236)
```
This time clustering results are stored in `Hm2236` and later when you rerun `draw()`, it directly uses the results that are saved.

**Comment by AlexZambon** [comment]
Do you have a function to extract the gene IDs from each k-means cluster?

**Comment by kmshort** [comment]
> Do you have a function to extract the gene IDs from each k-means cluster? I found some threads online but they are a little old so I was wondering if you built in the useful functionality into the package yet? Thanks for all your help.

I'd like to know this too @AlexZambon

**Comment by jokergoo** [comment]
You can do as follows:

```r
ht = Heatmap(..., row_km = 2)
ht = draw(ht)
rd = row_order(ht) # a list of two 
rownames(mat)[ rd[[1]] ]. # the gene ids in the first cluster
```

See explanations in https://jokergoo.github.io/ComplexHeatmap-reference/book/a-single-heatmap.html#get-orders-and-dendrograms-from-heatmap

**Comment by kmshort** [comment]
@jokergoo 
Many, many thanks. I looked through the book, but it was not obvious. As a suggestion? It would be handy if that could be accessed somewhere within the ht object itself. The gene names/labels are ht@row_names_param$labels and order of those labels in the Heatmap is ht@row_order I think. If there was something like ht@row_cluster_assignment it'd be directly accessible.
Fantastic package!

**Comment by jokergoo** [comment]
Hi,

The thing is in the following code:

```r
ht = Heatmap(..., row_km = 2)
ht = draw(ht)
```

`Heatmap()` is only a constructor function which has not applied clustering yet. It only puts all the data and configurations in the returned object `ht` in the first line. The clustering is only applied within `draw()` function and the object `ht` then is updated with the clustering results.

**Comment by kmshort** [comment]
@jokergoo 
Ahh okay, it didn't make sense to me because in an Rstudio RMD, if I call ht directly without draw(ht) I see the plot result. I guess whatever rstudio calls to render the image does the calculation? Anyway, this is what made it confusing to me because that process was 'in the background'. For what it's worth, it would be neat if the application of the clustering was done within the ht object itself. Then just let draw() do simple drawing.
thanks again.

**Comment by jokergoo** [comment]
Under Rmd, if you directly type `ht`, actually it executes `print(ht)` or `show(ht)` internally and `show()` function internally applies `draw()`. This is also what *ggplot2** does where `ggplot() + .. + ..` does not really make the plot. Directly printing the ggplot object internally calls `show()` function which makes the plot.

I cannot apply clustering within `Heatmap()` because as I said, `Heatmap()` is only a constructor function which just put all the data and parameters in. Also remember multiple heatmaps can be added, thus the clustering can only be applied when all the heatmap data are there.

Making heatmaps with Complexheamtap involves a list of functions and classes, and it is not like other heatmap functions such as `pheatmap()` or `heatmap.2()` which is just like a one-step process.

If you use Rmarkdown, you can explicitly code like:

```r
ht = Heatmap(...). # this does not draw anything
ht = draw(ht) # this draws the heatmap
row_order(ht) # this retrieves the row indices
```

**Comment by kmshort** [comment]
Great, thanks for the thorough explanation @jokergoo
Love your work!

**Comment by ftamiro** [comment]
>
> You can do as follows:
> 
> ```r
> ht = Heatmap(..., row_km = 2)
> ht = draw(ht)
> rd = row_order(ht) # a list of two 
> rownames(mat)[ rd[[1]] ]. # the gene ids in the first cluster
> ```
> 
> See explanations in https://jokergoo.github.io/ComplexHeatmap-reference/book/a-single-heatmap.html#get-orders-and-dendrograms-from-heatmap

Just a small note on this. Maybe it goes without saying, but for me, it was not obvious: I also need to extract gene IDs from a certain cluster obtained using kmeans. The order of clusters in my heatmap (top to bottom) is 2,1,5,6,4,3 (their labels on the heatmap). When I check
 
`rd <- row_order(ht)`
`rd`

I see lists with the same exact order as in my heatmap, the first is $2, last is $3 (both numbers are within backticks `, I do not know how to put them here without making it a code block).

From what I see in the heatmap, I am interested in the cluster labeled as 4. Problem is, when I used:

`rownames(mat)[ rd[[4]] ]`

I noticed that I did NOT obtain the genes I want (luckily, I had labeled some of them beforehand, so it was easy to notice that this list did not contain any genes I saw from the labels on the heatmap).

What solved this was to just put the number in quotation marks, like this:

`rownames(mat)[ rd[['4']] ]`

In this case, I obtained my actual cluster labeled as 4, whilst before, without the quotation mark, I was obtaining it by index, so in reality, I was getting genes from cluster#6, which is 4th in my order.

Maybe it is me being still too naive in bioinformatics and I can slip on things like that. But, I think some functions from inside the package that allows extracting clusters in a controlled way, making sure that the user gets immediately the proper cluster he intends by looking at the heatmap, would be of great help and avoid this kind of possible mistake.
